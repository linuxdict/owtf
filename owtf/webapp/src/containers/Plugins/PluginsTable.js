/*
 * Plugins table component
 * This component manages all the plugins data and renders it in the form of a table
 *
 */
import React from "react";
import PropTypes from "prop-types";
import { filter } from "fuzzaldrin-plus";
import { Table, Checkbox, Tooltip, Text } from "evergreen-ui";

export default class PluginsTable extends React.Component {
  constructor(props, context) {
    super(props, context);

    this.state = {
      selectedRows: [], //array of checked plugins IDs
      codeSearch: "", //plugin code filter query
      nameSearch: "", //plugin name filter query
      typeSearch: "", //plugin type filter query
      groupSearch: "", //plugin group filter query
      helpSearch: "" //plugin description filter query
    };
  }

  /**
   * Filter the plugins based on the code, name, type, group and description property
   * @param {array} plugins list of all plugins
   */
  handleTableFilter = plugins => {
    const globalSearch = this.props.globalSearch.trim();
    const codeSearch = this.state.codeSearch.trim();
    const nameSearch = this.state.nameSearch.trim();
    const typeSearch = this.state.typeSearch.trim();
    const groupSearch = this.state.groupSearch.trim();
    const helpSearch = this.state.helpSearch.trim();

    // If the searchQuery is empty, return the plugins as it is.
    if (
      codeSearch.length === 0 &&
      nameSearch.length === 0 &&
      typeSearch.length === 0 &&
      groupSearch.length === 0 &&
      helpSearch.length === 0 &&
      globalSearch.length === 0
    )
      return plugins;

    return plugins.filter(plugin => {
      // Use the filter from fuzzaldrin-plus to filter by url, method and status.
      var res = true,
        globalRes = true,
        resultCode,
        resultSearch,
        resultType,
        resultGroup,
        resultHelp;
      if (codeSearch.length) {
        resultCode = filter([plugin.code], codeSearch);
        res = res && resultCode.length === 1;
      }
      if (nameSearch.length) {
        resultSearch = filter([plugin.name.replace(/_/g, " ")], nameSearch);
        res = res && resultSearch.length === 1;
      }
      if (typeSearch.length) {
        resultType = filter([plugin.type.replace(/_/g, " ")], typeSearch);
        res = res && resultType.length === 1;
      }
      if (groupSearch.length) {
        resultGroup = filter([plugin.group], groupSearch);
        res = res && resultGroup.length === 1;
      }
      if (helpSearch.length) {
        resultHelp = filter([plugin.descrip], helpSearch);
        res = res && resultHelp.length === 1;
      }
      //Filter the worklist using the top search box
      if (globalSearch.length) {
        resultCode = filter([plugin.code], globalSearch);
        resultSearch = filter([plugin.name.replace(/_/g, " ")], globalSearch);
        resultType = filter([plugin.type.replace(/_/g, " ")], globalSearch);
        resultGroup = filter([plugin.group], globalSearch);
        resultHelp = filter([plugin.descrip], globalSearch);
        globalRes =
          resultCode.length === 1 ||
          resultSearch.length === 1 ||
          resultType.length === 1 ||
          resultGroup.length === 1 ||
          resultHelp.length === 1;
      }
      return res && globalRes;
    });
  };

  /**
   * Function updating the code filter qurey
   * @param {string} value code filter query
   */
  handleCodeFilterChange = value => {
    this.setState({ codeSearch: value });
  };

  /**
   * Function updating the name filter qurey
   * @param {string} value name filter query
   */
  handleNameFilterChange = value => {
    this.setState({ nameSearch: value });
  };

  /**
   * Function updating the type filter qurey
   * @param {string} value type filter query
   */
  handleTypeFilterChange = value => {
    this.setState({ typeSearch: value });
  };

  /**
   * Function updating the group filter qurey
   * @param {string} value group filter query
   */
  handleGroupFilterChange = value => {
    this.setState({ groupSearch: value });
  };

  /**
   * Function updating the description filter qurey
   * @param {string} value description filter query
   */
  handleHelpFilterChange = value => {
    this.setState({ helpSearch: value });
  };

  /**
   * Function updating the selected plugins after checking the select-all checkbox
   * @param {object} e checkbox onchange event
   * @param {array} plugins list of all plugins
   */
  handleSelectAllCheckBox = (e, plugins) => {
    const all_plugins = plugins.map(r => this.getPluginDetails(r));
    if (e.target.checked) {
      this.setState({ selectedRows: all_plugins }, () => {
        this.props.updateSelectedPlugins(this.state.selectedRows);
      });
    } else {
      this.setState({ selectedRows: [] }, () => {
        this.props.updateSelectedPlugins(this.state.selectedRows);
      });
    }
  };

  /**
   * Function updating the selected plugins after checking a row checkbox
   * @param {object} e checkbox onchange event
   * @param {object} plugin plugin corresponding to that checkbox
   */
  handleCheckBox = (e, plugin) => {
    const pluginDetails = this.getPluginDetails(plugin);
    if (e.target.checked) {
      this.setState(
        { selectedRows: [...this.state.selectedRows, pluginDetails] },
        () => {
          this.props.updateSelectedPlugins(this.state.selectedRows);
        }
      );
    } else {
      this.setState(
        {
          selectedRows: this.state.selectedRows.filter(
            x => x.key !== pluginDetails.key
          )
        },
        () => {
          this.props.updateSelectedPlugins(this.state.selectedRows);
        }
      );
    }
  };

  /**
   * Function to fetch details of plugin from a row
   * @param {object} plugin plugin corresponding to that row
   */
  getPluginDetails = plugin => {
    return {
      key: plugin.key,
      group: plugin.group,
      type: plugin.type,
      code: plugin.code
    };
  };

  render() {
    const items = this.handleTableFilter(this.props.plugins);
    return (
      <Table border>
        <Table.Head>
          <Table.HeaderCell flex={-1}>
            <Checkbox
              checked={this.state.selectedRows.length === items.length}
              onChange={e => this.handleSelectAllCheckBox(e, items)}
            />
          </Table.HeaderCell>
          <Table.SearchHeaderCell
            flex="none"
            width={150}
            onChange={this.handleCodeFilterChange}
            value={this.state.codeSearch}
            placeholder="Code"
          />
          <Table.SearchHeaderCell
            flex="none"
            width={200}
            onChange={this.handleNameFilterChange}
            value={this.state.nameSearch}
            placeholder="Name"
          />
          <Table.SearchHeaderCell
            flex="none"
            width={150}
            onChange={this.handleTypeFilterChange}
            value={this.state.typeSearch}
            placeholder="Type"
          />
          <Table.SearchHeaderCell
            flex="none"
            width={150}
            onChange={this.handleGroupFilterChange}
            value={this.state.groupSearch}
            placeholder="Group"
          />
          <Table.SearchHeaderCell
            flex="none"
            width={200}
            onChange={this.handleHelpFilterChange}
            value={this.state.helpSearch}
            placeholder="Help"
          />
        </Table.Head>
        <Table.VirtualBody height={500}>
          {items.map(plugin => (
            <Table.Row key={plugin.key} height="auto" isSelectable>
              <Table.Cell flex={-1}>
                <Checkbox
                  checked={
                    this.state.selectedRows.filter(p => p.key === plugin.key)
                      .length > 0
                  }
                  id={plugin.key}
                  onChange={e => this.handleCheckBox(e, plugin)}
                />
              </Table.Cell>
              <Table.TextCell flex="none" width={150}>
                {plugin.code}
              </Table.TextCell>
              <Table.TextCell flex="none" width={200}>
                {plugin.name.replace(/_/g, " ")}
              </Table.TextCell>
              <Table.TextCell flex="none" width={150}>
                {plugin.type.replace(/_/g, " ")}
              </Table.TextCell>
              <Table.TextCell flex="none" width={150}>
                {plugin.group}
              </Table.TextCell>
              <Table.TextCell flex="none" width={200}>
                <Tooltip content={plugin.descrip}>
                  <Text>{plugin.descrip}</Text>
                </Tooltip>
              </Table.TextCell>
            </Table.Row>
          ))}
        </Table.VirtualBody>
      </Table>
    );
  }
}

PluginsTable.propTypes = {
  plugins: PropTypes.array,
  globalSearch: PropTypes.string,
  updateSelectedPlugins: PropTypes.func
};
