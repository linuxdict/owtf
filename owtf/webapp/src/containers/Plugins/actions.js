/*
 * Plugin Actions
 *
 * Actions change things in your application
 * Since this owtf uses a uni-directional data flow, specifically redux,
 * we have these actions which are the only way your application interacts with
 * your application state. This guarantees that your state is up to date and nobody
 * messes it up weirdly somewhere.
 *
 * To add a new Action:
 * 1) Import your constant
 * 2) Add a function like this:
 *    export function yourAction(var) {
 *        return { type: YOUR_ACTION_CONSTANT, var: var }
 *    }
 */

import {
    LOAD_PLUGINS,
    LOAD_PLUGINS_SUCCESS,
    LOAD_PLUGINS_ERROR,
    POST_TO_WORKLIST,
    POST_TO_WORKLIST_SUCCESS,
    POST_TO_WORKLIST_ERROR,
  } from './constants';
  
  /**
   * Load the plugins, this action starts the request saga GET
   *
   * @return {object} An action object with a type of LOAD_PLUGINS
   */
  export function loadPlugins() {
    return {
      type: LOAD_PLUGINS,
    };
  }
  
  /**
   * Dispatched when the plugins are loaded by the request saga
   *
   * @param  {array} plugins The plugins data
   *
   * @return {object} An action object with a type of LOAD_PLUGINS_SUCCESS passing the plugins
   */
  export function pluginsLoaded(plugins) {
    return {
      type: LOAD_PLUGINS_SUCCESS,
      plugins,
    };
  }
  
  /**
   * Dispatched when loading the plugins fails
   *
   * @param  {object} error The error
   *
   * @return {object} An action object with a type of LOAD_PLUGINS_ERROR passing the error
   */
  export function pluginsLoadingError(error) {
    return {
      type: LOAD_PLUGINS_ERROR,
      error,
    };
  }

  /**
   * Post the targets to worklist, this action starts the request saga POST.
   *
   * @param  {string} plugin_data data of the launched plugins.
   *
   * @return {object} An action object with a type of CREATE_TARGET
   */
  export function postToWorklist(plugin_data) {
    return {
      type: POST_TO_WORKLIST,
      plugin_data
    };
  }
  
  /**
   * Dispatched when the targets are posted to the worklist by the request saga
   *
   * @return {object} An action object with a type of CREATE_TARGET_SUCCESS
   */
  export function targetPosted() {
    return {
      type: POST_TO_WORKLIST_SUCCESS,
    };
  }
  
  /**
   * Dispatched when posting the target fails
   *
   * @param  {object} error The error
   *
   * @return {object} An action object with a type of CREATE_TARGET_ERROR passing the error
   */
  export function targetPostingError(error) {
    return {
      type: POST_TO_WORKLIST_ERROR,
      error,
    };
  }