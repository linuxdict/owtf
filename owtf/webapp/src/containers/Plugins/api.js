import Request from "../../utils/request";
import { API_BASE_URL } from "../../utils/constants";

function getHeaders() {
  return {
    headers: {
      "Content-Type": "application/x-www-form-urlencoded;charset=UTF-8",
      Authorization: `Bearer ${localStorage.getItem("token")}`
    }
  };
}

export function getPluginsAPI() {
  const requestURL = `${API_BASE_URL}plugins/`;
  const options = getHeaders();
  const request = new Request(requestURL, options);
  return request.get.bind(request);
}

export function postTargetsToWorklistAPI() {
  const requestURL = `${API_BASE_URL}worklist/`;
  const options = getHeaders();
  const request = new Request(requestURL, options);
  return request.post.bind(request);
}
