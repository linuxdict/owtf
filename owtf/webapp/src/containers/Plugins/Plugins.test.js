import React from "react";
import { shallow } from "enzyme";
import toJson from "enzyme-to-json";
import "../../setupTests";
import { checkProps } from "../../utils/testUtils";
import configureStore from "redux-mock-store";
import ConnectedPlugins, { Plugins } from "./index";
import PluginsTable from "./PluginsTable";
import { fromJS } from "immutable";

import {
  pluginsLoaded,
  pluginsLoadingError,
  targetPosted,
  targetPostingError
} from "./actions";
import {
  LOAD_PLUGINS,
  LOAD_PLUGINS_SUCCESS,
  LOAD_PLUGINS_ERROR,
  POST_TO_WORKLIST,
  POST_TO_WORKLIST_SUCCESS,
  POST_TO_WORKLIST_ERROR
} from "./constants";

import { getPlugins, postTargetsToWorklist } from "./saga";
import { runSaga } from "redux-saga";
import * as api from "./api";

import { pluginsLoadReducer, postToWorklistReducer } from "./reducer";
import { loadTargets } from "../TargetsPage/actions";

const setUp = (initialState = {}) => {
  const mockStore = configureStore();
  const store = mockStore(initialState);
  const wrapper = shallow(<ConnectedPlugins store={store} />);
  return wrapper;
};

describe("Plugins componemt", () => {
  describe("Testing dumb plugins component", () => {
    let wrapper;
    let props;
    beforeEach(() => {
      props = {
        loading: false,
        error: false,
        plugins: [
          {
            key: "semi_passive@OWTF-IG-004",
            group: "web",
            name: "Web_Application_Fingerprint",
            code: "OWTF-IG-004",
            descrip: "Normal requests to gather fingerprint info",
            type: "semi_passive"
          }
        ],
        postingError: false,
        pluginShow: false,
        onFetchPlugins: jest.fn(),
        onPostToWorklist: jest.fn(),
        handlePluginClose: jest.fn(),
        selectedTargets: [],
        handleAlertMsg: jest.fn(),
        resetTargetState: jest.fn()
      };
      wrapper = shallow(<Plugins {...props} />);
    });

    it("Should have correct prop-types", () => {
      const expectedProps = {
        loading: false,
        error: false,
        plugins: [],
        postingError: false,
        pluginShow: true,
        onFetchPlugins: () => {},
        onPostToWorklist: () => {},
        handlePluginClose: () => {},
        selectedTargets: [1, 2],
        handleAlertMsg: () => {},
        resetTargetState: () => {}
      };
      const propsErr = checkProps(Plugins, expectedProps);
      expect(propsErr).toBeUndefined();
    });

    it("Should render without errors", () => {
      const component = wrapper.find("withTheme(Dialog)");
      expect(component.length).toBe(1);
      expect(toJson(component)).toMatchSnapshot();
    });

    it("Should correctly render its sub-components", () => {
      const tablist = wrapper.find("Tablist");
      const tabs = wrapper.find("withTheme(Tab)");
      const individualPanel = wrapper.find("#panel-individual");
      const groupPanel = wrapper.find("#panel-group");
      const checkbox = wrapper.find("#force-overwrite");
      const pluginTable = wrapper.find("PluginsTable");

      expect(tablist.length).toBe(1);
      expect(tabs.length).toBe(3);
      expect(tabs.at(0).props().children).toEqual("Launch Individually");
      expect(tabs.at(1).props().children).toEqual("Launch in groups");
      expect(individualPanel.length).toBe(1);
      expect(groupPanel.length).toBe(1);
      expect(checkbox.length).toBe(1);
      expect(pluginTable.length).toBe(1);
    });

    it("Should properly switch panels on tab click", () => {
      const individualTab = wrapper.find("withTheme(Tab)").at(0);
      const groupTab = wrapper.find("withTheme(Tab)").at(1);
      groupTab.simulate("select");
      expect(wrapper.instance().state.selectedIndex).toEqual(2);
      individualTab.simulate("select");
      expect(wrapper.instance().state.selectedIndex).toEqual(1);
    });

    it("handleGroupLaunch should correcly return plugins group & type array", () => {
      const expectedGroupArray = [["web"], ["semi_passive"]];
      const plugins = wrapper.instance();
      const groupArray = plugins.handleGroupLaunch();
      expect(groupArray).toEqual(expectedGroupArray);
    });

    it("Should pass correct props to its child components", () => {
      const pluginTable = wrapper.find("PluginsTable");
      expect(pluginTable.props().plugins).toEqual(props.plugins);
    });
  });

  describe("Testing connected Plugins component", () => {
    let wrapper;
    let initialState;
    beforeEach(() => {
      const load = {
        loading: true,
        error: false,
        plugins: false
      };
      const postToWorklist = {
        loading: false,
        error: {}
      };
      const plugins = {
        load,
        postToWorklist
      };
      initialState = fromJS({
        plugins
      });
      wrapper = setUp(initialState);
    });

    it("Props should match the initial state", () => {
      const pluginsProp = initialState
        .get("plugins")
        .get("load")
        .get("plugins");
      const fetchErrorProp = initialState
        .get("plugins")
        .get("load")
        .get("error");
      const fetchLoadingProp = initialState
        .get("plugins")
        .get("load")
        .get("loading");
      const createErrorProp = initialState
        .get("plugins")
        .get("postToWorklist")
        .get("error");

      expect(wrapper.props().plugins).toEqual(pluginsProp);
      expect(wrapper.props().loading).toEqual(fetchLoadingProp);
      expect(wrapper.props().error).toEqual(fetchErrorProp);
      expect(wrapper.props().postingError).toEqual(createErrorProp);
    });
  });

  describe("Testing Plugins Table component", () => {
    let wrapper;
    let props;
    beforeEach(() => {
      props = {
        plugins: [
          {
            key: "semi_passive@OWTF-IG-004",
            group: "web",
            name: "Web_Application_Fingerprint",
            code: "OWTF-IG-004",
            descrip: "Normal requests to gather fingerprint info",
            type: "semi_passive"
          },
          {
            key: "dos@OWTF-ADoS-001",
            group: "auxiliary",
            name: "Direct_DoS_Launcher",
            code: "OWTF-ADoS-001",
            descrip: "Denial of Service (DoS) Launcher",
            type: "dos"
          }
        ],
        updateSelectedPlugins: jest.fn(),
        globalSearch: ""
      };
      wrapper = shallow(<PluginsTable {...props} />);
    });

    it("Should have correct prop-types", () => {
      const expectedProps = {
        plugins: [],
        updateSelectedPlugins: () => {}
      };
      const propsErr = checkProps(PluginsTable, expectedProps);
      expect(propsErr).toBeUndefined();
    });

    it("Should render without errors", () => {
      const component = wrapper.find("Table");
      expect(component.length).toBe(1);
      expect(toJson(component)).toMatchSnapshot();
    });

    it("Should correctly render table's sub-components", () => {
      const searchHeader = wrapper.find("SearchTableHeaderCell");
      const checkBoxHeader = wrapper.find("TableHeaderCell");
      const tableRow = wrapper.find("withTheme(TableRow)");
      const textCell = wrapper.find("TextTableCell");
      const cell = wrapper.find("withTheme(TableCell)");
      const checkBox = wrapper.find("withTheme(Checkbox)");

      expect(searchHeader.length).toBe(5);
      expect(searchHeader.at(0).props().placeholder).toEqual("Code");
      expect(searchHeader.at(1).props().placeholder).toEqual("Name");
      expect(searchHeader.at(2).props().placeholder).toEqual("Type");
      expect(searchHeader.at(3).props().placeholder).toEqual("Group");
      expect(searchHeader.at(4).props().placeholder).toEqual("Help");
      expect(checkBoxHeader.length).toBe(1);
      expect(tableRow.length).toBe(props.plugins.length);
      expect(textCell.length).toBe(5 * props.plugins.length);
      expect(textCell.at(0).props().children).toEqual(props.plugins[0].code);
      expect(textCell.at(3).props().children).toEqual(props.plugins[0].group);
      expect(cell.length).toBe(props.plugins.length);
      expect(checkBox.length).toBe(1 + props.plugins.length);
    });

    it("Should call updateSelectedPlugins on checkbox click", () => {
      expect(props.updateSelectedPlugins.mock.calls.length).toBe(0);
      const checkBox = wrapper.find("withTheme(Checkbox)").at(0);
      const event = {
        preventDefault() {},
        target: { checked: false }
      };
      checkBox.simulate("change", event);
      expect(props.updateSelectedPlugins.mock.calls.length).toBe(1);
    });

    it("Should filter the plugins correctly", () => {
      wrapper.setState({ codeSearch: "IG" });
      expect(wrapper.find("withTheme(TableRow)").length).toBe(1);
      expect(
        wrapper
          .find("TextTableCell")
          .at(0)
          .props().children
      ).toEqual("OWTF-IG-004");
      wrapper.setState({ codeSearch: "", nameSearch: "DoS" });
      expect(wrapper.find("withTheme(TableRow)").length).toBe(1);
      wrapper.setState({
        nameSearch: "",
        typeSearch: "dos",
        groupSearch: "web"
      });
      expect(wrapper.find("withTheme(TableRow)").length).toBe(0);
      wrapper.setState({
        helpSearch: "Denial",
        typeSearch: "",
        groupSearch: ""
      });
      expect(wrapper.find("withTheme(TableRow)").length).toBe(1);
      expect(
        wrapper
          .find("TextTableCell")
          .at(0)
          .props().children
      ).toEqual("OWTF-ADoS-001");
    });
  });

  describe("Testing the sagas", () => {
    describe("Testing getPlugins saga", () => {
      it("Should load plugins in case of success", async () => {
        // we push all dispatched actions to make assertions easier
        // and our tests less brittle
        const dispatchedActions = [];

        // we don't want to perform an actual api call in our tests
        // so we will mock the fetchAPI api with jest
        // this will mutate the dependency which we may reset if other tests
        // are dependent on it
        const mockedPlugins = {
          status: "success",
          data: [
            {
              key: "semi_passive@OWTF-IG-004",
              group: "web",
              name: "Web_Application_Fingerprint",
              code: "OWTF-IG-004",
              descrip: "Normal requests to gather fingerprint info",
              type: "semi_passive"
            }
          ]
        };
        api.getPluginsAPI = jest.fn(() =>
          jest.fn(() => Promise.resolve(mockedPlugins))
        );

        const fakeStore = {
          getState: () => ({ state: "test" }),
          dispatch: action => dispatchedActions.push(action)
        };

        // wait for saga to complete
        await runSaga(fakeStore, getPlugins).done;

        expect(api.getPluginsAPI.mock.calls.length).toBe(1);
        expect(dispatchedActions).toContainEqual(
          pluginsLoaded(mockedPlugins.data)
        );
      });

      it("Should handle plugins load errors in case of failure", async () => {
        const dispatchedActions = [];

        // we simulate an error by rejecting the promise
        // then we assert if our saga dispatched the action(s) correctly
        const error = "API server is down";
        api.getPluginsAPI = jest.fn(() => jest.fn(() => Promise.reject(error)));

        const fakeStore = {
          getState: () => ({ state: "test" }),
          dispatch: action => dispatchedActions.push(action)
        };

        await runSaga(fakeStore, getPlugins).done;

        expect(api.getPluginsAPI.mock.calls.length).toBe(1);
        expect(dispatchedActions).toContainEqual(pluginsLoadingError(error));
      });
    });

    describe("Testing postTargetsToWorklist saga", () => {
      let fakeAction;
      beforeEach(() => {
        fakeAction = {
          type: POST_TO_WORKLIST,
          plugin_data: {}
        };
      });

      it("Should post selected targets to worklist and load targets in case of success", async () => {
        const dispatchedActions = [];
        api.postTargetsToWorklistAPI = jest.fn(() =>
          jest.fn(() => Promise.resolve())
        );

        const fakeStore = {
          getState: () => ({ state: "test" }),
          dispatch: action => dispatchedActions.push(action)
        };
        await runSaga(fakeStore, postTargetsToWorklist, fakeAction).done;

        expect(api.postTargetsToWorklistAPI.mock.calls.length).toBe(1);
        expect(dispatchedActions).toContainEqual(targetPosted());
        expect(dispatchedActions).toContainEqual(loadTargets());
      });

      it("Should handle target posting error in case of failure", async () => {
        const dispatchedActions = [];
        const error = "API server is down";
        api.postTargetsToWorklistAPI = jest.fn(() =>
          jest.fn(() => Promise.reject(error))
        );

        const fakeStore = {
          getState: () => ({ state: "test" }),
          dispatch: action => dispatchedActions.push(action)
        };

        await runSaga(fakeStore, postTargetsToWorklist, fakeAction).done;

        expect(api.postTargetsToWorklistAPI.mock.calls.length).toBe(1);
        expect(dispatchedActions).toContainEqual(targetPostingError(error));
      });
    });
  });

  describe("Testing reducers", () => {
    describe("Testing pluginsLoadReducer", () => {
      let initialState;
      beforeEach(() => {
        initialState = {
          loading: true,
          error: false,
          plugins: false
        };
      });

      it("Should return the initial state", () => {
        const newState = pluginsLoadReducer(undefined, {});
        expect(newState.toJS()).toEqual(initialState);
      });

      it("Should handle LOAD_PLUGINS", () => {
        const newState = pluginsLoadReducer(undefined, {
          type: LOAD_PLUGINS
        });
        expect(newState.toJS()).toEqual({
          loading: true,
          error: false,
          plugins: false
        });
      });

      it("Should handle LOAD_PLUGINS_SUCCESS", () => {
        const plugins = [
          {
            key: "semi_passive@OWTF-IG-004",
            group: "web",
            name: "Web_Application_Fingerprint",
            code: "OWTF-IG-004",
            descrip: "Normal requests to gather fingerprint info",
            type: "semi_passive"
          }
        ];
        const newState = pluginsLoadReducer(
          fromJS({
            loading: true,
            error: true,
            plugins: false
          }),
          {
            type: LOAD_PLUGINS_SUCCESS,
            plugins
          }
        );
        expect(newState.toJS()).toEqual({
          loading: false,
          error: false,
          plugins: plugins
        });
      });

      it("Should handle LOAD_PLUGINS_ERROR", () => {
        const newState = pluginsLoadReducer(
          fromJS({
            loading: true,
            error: true,
            plugins: false
          }),
          {
            type: LOAD_PLUGINS_ERROR,
            error: "Test plugins loading error"
          }
        );
        expect(newState.toJS()).toEqual({
          loading: false,
          error: "Test plugins loading error",
          plugins: false
        });
      });
    });

    describe("Testing postToWorklistReducer", () => {
      let initialState;
      beforeEach(() => {
        initialState = {
          loading: false,
          error: false
        };
      });

      it("Should return the initial state", () => {
        const newState = postToWorklistReducer(undefined, {});
        expect(newState.toJS()).toEqual(initialState);
      });

      it("Should handle POST_TO_WORKLIST,", () => {
        const newState = postToWorklistReducer(undefined, {
          type: POST_TO_WORKLIST,
          plugin_data: {}
        });
        expect(newState.toJS()).toEqual({ loading: true, error: false });
      });

      it("Should handle POST_TO_WORKLIST,_SUCCESS", () => {
        const newState = postToWorklistReducer(
          fromJS({
            loading: true,
            error: true
          }),
          {
            type: POST_TO_WORKLIST_SUCCESS
          }
        );
        expect(newState.toJS()).toEqual({
          loading: false,
          error: false
        });
      });

      it("Should handle POST_TO_WORKLIST,_ERROR", () => {
        const newState = postToWorklistReducer(
          fromJS({
            loading: true,
            error: true
          }),
          {
            type: POST_TO_WORKLIST_ERROR,
            error: "Test target posting error"
          }
        );
        expect(newState.toJS()).toEqual({
          loading: false,
          error: "Test target posting error"
        });
      });
    });
  });
});
