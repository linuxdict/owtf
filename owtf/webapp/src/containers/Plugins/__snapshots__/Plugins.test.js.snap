// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Plugins componemt Testing Plugins Table component Should render without errors 1`] = `
<Table
  border={true}
>
  <TableHead
    accountForScrollbar={true}
    height={32}
  >
    <TableHeaderCell
      flex={-1}
    >
      <withTheme(Checkbox)
        checked={false}
        onChange={[Function]}
      />
    </TableHeaderCell>
    <SearchTableHeaderCell
      flex="none"
      icon="search"
      onChange={[Function]}
      placeholder="Code"
      spellCheck={true}
      value=""
      width={150}
    />
    <SearchTableHeaderCell
      flex="none"
      icon="search"
      onChange={[Function]}
      placeholder="Name"
      spellCheck={true}
      value=""
      width={200}
    />
    <SearchTableHeaderCell
      flex="none"
      icon="search"
      onChange={[Function]}
      placeholder="Type"
      spellCheck={true}
      value=""
      width={150}
    />
    <SearchTableHeaderCell
      flex="none"
      icon="search"
      onChange={[Function]}
      placeholder="Group"
      spellCheck={true}
      value=""
      width={150}
    />
    <SearchTableHeaderCell
      flex="none"
      icon="search"
      onChange={[Function]}
      placeholder="Help"
      spellCheck={true}
      value=""
      width={200}
    />
  </TableHead>
  <TableVirtualBody
    allowAutoHeight={false}
    defaultHeight={48}
    height={500}
    overscanCount={5}
    useAverageAutoHeightEstimation={true}
  >
    <withTheme(TableRow)
      height="auto"
      isSelectable={true}
      key="semi_passive@OWTF-IG-004"
    >
      <withTheme(TableCell)
        flex={-1}
      >
        <withTheme(Checkbox)
          checked={false}
          id="semi_passive@OWTF-IG-004"
          onChange={[Function]}
        />
      </withTheme(TableCell)>
      <TextTableCell
        flex="none"
        isNumber={false}
        width={150}
      >
        OWTF-IG-004
      </TextTableCell>
      <TextTableCell
        flex="none"
        isNumber={false}
        width={200}
      >
        Web Application Fingerprint
      </TextTableCell>
      <TextTableCell
        flex="none"
        isNumber={false}
        width={150}
      >
        semi passive
      </TextTableCell>
      <TextTableCell
        flex="none"
        isNumber={false}
        width={150}
      >
        web
      </TextTableCell>
      <TextTableCell
        flex="none"
        isNumber={false}
        width={200}
      >
        <Tooltip
          appearance="default"
          content="Normal requests to gather fingerprint info"
          hideDelay={120}
          position="bottom"
          showDelay={0}
        >
          <withTheme(Text)>
            Normal requests to gather fingerprint info
          </withTheme(Text)>
        </Tooltip>
      </TextTableCell>
    </withTheme(TableRow)>
    <withTheme(TableRow)
      height="auto"
      isSelectable={true}
      key="dos@OWTF-ADoS-001"
    >
      <withTheme(TableCell)
        flex={-1}
      >
        <withTheme(Checkbox)
          checked={false}
          id="dos@OWTF-ADoS-001"
          onChange={[Function]}
        />
      </withTheme(TableCell)>
      <TextTableCell
        flex="none"
        isNumber={false}
        width={150}
      >
        OWTF-ADoS-001
      </TextTableCell>
      <TextTableCell
        flex="none"
        isNumber={false}
        width={200}
      >
        Direct DoS Launcher
      </TextTableCell>
      <TextTableCell
        flex="none"
        isNumber={false}
        width={150}
      >
        dos
      </TextTableCell>
      <TextTableCell
        flex="none"
        isNumber={false}
        width={150}
      >
        auxiliary
      </TextTableCell>
      <TextTableCell
        flex="none"
        isNumber={false}
        width={200}
      >
        <Tooltip
          appearance="default"
          content="Denial of Service (DoS) Launcher"
          hideDelay={120}
          position="bottom"
          showDelay={0}
        >
          <withTheme(Text)>
            Denial of Service (DoS) Launcher
          </withTheme(Text)>
        </Tooltip>
      </TextTableCell>
    </withTheme(TableRow)>
  </TableVirtualBody>
</Table>
`;

exports[`Plugins componemt Testing dumb plugins component Should render without errors 1`] = `
<withTheme(Dialog)
  confirmLabel="Run"
  isShown={false}
  onCloseComplete={[MockFunction]}
  onConfirm={[Function]}
  title="Plugins"
  width={1000}
>
  <withTheme(Pane)
    alignItems="center"
    display="flex"
    flexDirection="row"
  >
    <Tablist
      flexBasis={800}
      marginBottom={16}
      marginRight={24}
    >
      <withTheme(Tab)
        aria-controls="panel-individual"
        id={1}
        isSelected={true}
        key="1"
        onSelect={[Function]}
      >
        Launch Individually
      </withTheme(Tab)>
      <withTheme(Tab)
        aria-controls="panel-group"
        id={2}
        isSelected={false}
        key="2"
        onSelect={[Function]}
      >
        Launch in groups
      </withTheme(Tab)>
      <withTheme(Tab)>
        <withTheme(SearchInput)
          borderRadius={100}
          className="search-box"
          flex={1}
          onChange={[Function]}
          placeholder="Search"
          value=""
        />
      </withTheme(Tab)>
    </Tablist>
    <withTheme(Checkbox)
      checked={false}
      className="pull-right"
      id="force-overwrite"
      label="Force Overwrite"
      onChange={[Function]}
    />
  </withTheme(Pane)>
  <withTheme(Pane)
    flex="1"
    padding={16}
  >
    <withTheme(Pane)
      aria-hidden={false}
      aria-labelledby="individual"
      display="block"
      id="panel-individual"
      key="1"
      role="tabpanel"
    >
      <PluginsTable
        globalSearch=""
        plugins={
          Array [
            Object {
              "code": "OWTF-IG-004",
              "descrip": "Normal requests to gather fingerprint info",
              "group": "web",
              "key": "semi_passive@OWTF-IG-004",
              "name": "Web_Application_Fingerprint",
              "type": "semi_passive",
            },
          ]
        }
        updateSelectedPlugins={[Function]}
      />
    </withTheme(Pane)>
    <withTheme(Pane)
      aria-hidden={true}
      aria-labelledby="group"
      display="none"
      id="panel-group"
      key="2"
      role="tabpanel"
    >
      <withTheme(Pane)
        display="flex"
        flexDirection="row"
      >
        <withTheme(Pane)
          display="flex"
          flexDirection="column"
          width={600}
        >
          <withTheme(Heading)
            marginTop="default"
          >
            Plugin Groups
          </withTheme(Heading)>
          <withTheme(Checkbox)
            checked={false}
            key="0"
            label="web"
            onChange={[Function]}
          />
        </withTheme(Pane)>
        <withTheme(Pane)
          display="flex"
          flexDirection="column"
        >
          <withTheme(Heading)
            marginTop="default"
          >
            Plugin Types
          </withTheme(Heading)>
          <withTheme(Checkbox)
            checked={false}
            key="0"
            label="semi passive"
            onChange={[Function]}
          />
        </withTheme(Pane)>
      </withTheme(Pane)>
    </withTheme(Pane)>
  </withTheme(Pane)>
</withTheme(Dialog)>
`;
